File: 9. Build BST.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


class BST:
    def __init__(self):
        self.root = None

    def insert(self, data):
        if self.root is None:
            self.root = Node(data)
        else:
            self.insertRecursion(self.root, data)

    def insertRecursion(self, node, data):
        if data < node.data:
            if node.left is None:
                node.left = Node(data)
            else:
                self.insertRecursion(node.left, data)
        else:
            if node.right is None:
                node.right = Node(data)
            else:
                self.insertRecursion(node.right, data)

    def inorder(self, node):
        if node is None:
            return
        self.inorder(node.left)
        print(node.data, end=" ")
        self.inorder(node.right)


line = map(int,input().split())
tree = BST()
for x in line:
    if x == -1:
        break
    tree.insert(x)

tree.inorder(tree.root)

File: 9. Build BST.java
import java.util.*;
class Node 
{
    int data;
    Node left, right;
    Node(int data) 
    {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

class BST 
{
    Node root;
    public void insert(int data) 
    {
        if (root == null)  root = new Node(data);
        else  insertRecursion(root, data);
    }

    public void insertRecursion(Node node, int data) 
    {
        if (data < node.data) 
        {
            if (node.left == null) 
            {
                node.left = new Node(data);
            } 
            else 
            {
                insertRecursion(node.left, data);
            }
        } 
        else 
        {
            if (node.right == null) 
            {
                node.right = new Node(data);
            } 
            else 
            {
                insertRecursion(node.right, data);
            }
        }
    }

    public void inOrder(Node node) 
    {
        if (node == null)
            return;
        inOrder(node.left);
        System.out.print(node.data + " ");
        inOrder(node.right);
    }
}

public class Main {
    public static void main(String[] args) 
    {
        Scanner sc = new Scanner(System.in);
        String line = sc.nextLine();         
        String[] tokens = line.split(" ");   
        BST tree = new BST();

        for (String token : tokens) 
        {
            int val = Integer.valueOf(token);
            if (val == -1)
                break;

            tree.insert(val);
        }

        tree.inOrder(tree.root);
    }
}
